docs: Added tutorial topic

Doug-AWS opened the pull request:
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

Resolves comments from PR #872; Steve's review.

rix0rrr commented on docs/src/tutorial.rst:
```diff
@@ -0,0 +1,496 @@
+    exports.main = function(event, context, callback) {
+      if (event.httpMethod != null) {
+        if (event.httpMethod != "") {
+          if (event.httpMethod == "GET") {
```
These three lines below each other make me restless. There should be a better way to express this.

Doug-AWS replied:
I've refactored the code to use a try block.

rix0rrr replied:
I don't see how that helps.

What's making me nervous here, and I should have been more clear, is you're doing the following:

```
if (variable unequal to one thing)
AND
if (variable unequal to other thing)
AND
if (variable IS EUQAL to something else)
THEN do something
```

Well first of all, if `event.httpMethod == "GET"`, then of course it must be `!= null` and `!= ""`, so there's no need to check those. You could have gotten away with writing:

```ts
if (event.httpMethod == "GET") {
   // ...
}
if (event.httpMethod == "POST") {
  // ...
}
// etc
```
And have the same effect.

The only reason you would type it this way, is because maybe at the end you might write an `else`, something like this:

```ts
if (event.httpMethod != "") {
  if (Event.httpMethod == "GET") {
   // lots of
  }
  // ...lines here...
} else {
  // Do something special
}
```

However, because the inner block is SO LONG, I have a hard time visually scanning whether that's true or not. So I recommend you invert the condition into an early abort, if that's the case:

```ts
if (event.httpMethod == "") {
  // Do something special
  return;
}
if (event.httpMethod == "GET") {
   /// ...
}
if (event.httpMethod == "POST") {
  // ...
}
// etc
```

Long blocks and redundant ifs make me nervous because they might be hiding logic that's easy to miss, and so I recommend that you change the code layout to be very clear and straightforward about what's happening.

I'm not sure how adding a `try` helps there. But also, I can't find the code back in the PR after it has exploded through the merge, so I'll have a look again after it has normalized.

-----

## Also

Forgot to take note of this first time around, but since you're doing JavaScript, REPLACE ALL INSTANCES of the following:

* `==` => `===`
* `!=` => `!==`

The 2-character variants are not to be trusted.

rix0rrr commented on docs/src/tutorial.rst:
```diff
@@ -0,0 +1,496 @@
+The next step is to create a |LAM| function to list all of the widgets in our
+|S3| bucket.
+
+Create the directory *resource* at the same level as the *bin* directory.
```
resource or resources?

rix0rrr replied:
Also is that the best name for this directory?

rix0rrr commented on docs/src/tutorial.rst:
```diff
@@ -0,0 +1,496 @@
+    exports.main = function(event, context, callback) {
+      if (event.httpMethod == "GET") {
+        if (event.path.length == 1) {
+          S3.listObjectsV2({ Bucket: bucketName })
```
Use NodeJS8 and you can rewrite this to be `async`, which is way more readable:

```ts
exports.main = async function(event, context) {
  try {
    if (event.httpMethod == "GET") {
      if (event.path.length == 1) {
        const data = await S3.listObjectsV2({ Bucket: bucketName }).promise();
        var body = { widgets: data.Contents.map(function(e) { return e.Key }) };
        return {
          statusCode: 200,
          headers: {},
          body: JSON.stringify(body)
        };
      }
    }
  } catch(error) {
    var body = error.stack || JSON.stringify(error, null, 2);
    var response = makeError(body);
    return response;
  }
}
```

rix0rrr replied:
By the way where does `makeError` come from?

Doug-AWS replied:
Okay, I no longer use helper functions.

rix0rrr replied:
That's not the point. I like helper functions. I just didn't see it defined. Or did I miss it?

rix0rrr commented on docs/src/tutorial.rst:
```diff
@@ -0,0 +1,496 @@
+
+    npm install @aws-cdk/aws-apigateway
+    npm install @aws-cdk/aws-lambda
+    npm install @aws-cdk/aws-s3
```
This is probably all possible on one line

Doug-AWS replied:
yep

rix0rrr commented on docs/src/tutorial.rst:
```diff
@@ -0,0 +1,496 @@
+.. code-block:: sh
+
+    npm run build
+    cdk synth
```
What's the point of this?

Doug-AWS replied:
I want to reinforce that when you make changes you should rebuild and see if they are what you want.

rix0rrr replied:
Fair enough

rix0rrr commented:
Something went a little weird with your merge.

Doug-AWS commented:
WTF is going on? When I had my dougsch/new-tutorial branch checked out, I ran "git diff --name-only master" and it gave me index.rst and tutorial.rst. Now it says I have 191 file diffs?

OK, I just rebased my branch against master:

git checkout master
git pull
git checkout dougsch/new-tutorial
git pull
git fetch --all -p
git fetch origin
git rebase origin/master

Got a merge conflict in tutorial.rst. How could that happen as master does not have a tutorial.rst????
Whatever, fixed the merge conflict, then ran:

git pull
git push
git status

Gave me:

On branch dougsch/new-tutorial
Your branch is up to date with 'origin/dougsch/new-tutorial'.

nothing to commit, working tree clean

git diff --name-only master

Gave me:

docs/src/index.rst
docs/src/tutorial.rst

What else am I supposed to do???

Doug-AWS commented:
Forget this. I'm trying a new branch, dougsch/tutorial, and new PR.

rix0rrr commented:
Doug, I think the problem is in your rebasing.

### Don't rebase!

Rebasing will try to move the commits around, change their identity, resurface merge conflicts that you've already solved, etc.

Rebasing is a good solution if you're committing directly to master and you want to keep your history clean. However, we work via PRs that we squash, so there's no need to keep the history clean. Any change goes into master as a single commit, and the history leading up to that commit is lost. Your life is going to be a lot better if you stick to the following:

```
# Start a new branch
(master)$ git checkout master
(master)$ git checkout -b dougsch/some-change
(dougsch/some-change)$ 

... work ...

# Ready to create a PR
(dougsch/some-change)$ git add -A .
(dougsch/some-change)$ git commit
(dougsch/some-change)$ git push -u

# Switch to another branch to continue working there
(dougsch/some-change)$ git checkout dougsch/oldpr
(dougsch/oldpr)$ git fetch 
# NO REBASE HERE!
(dougsch/oldpr)$ git merge origin/master

... work ...

# Update the PR branch
(dougsch/oldpr)$ git commit -a
(dougsch/oldpr)$ git push
```

If you don't yet have a script installed yet that shows your current branch in your shell prompt, get one today. Super handy!
